pkg/controllers/generatorstate/generatorstate_controller.go

```diff
@@ -0,0 +1,143 @@
+	err = r.Get(ctx, req.NamespacedName, generatorState)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, logErrorGetGS)
+		return ctrl.Result{}, err
+	}
+
+	requeue, err := r.handleFinalizer(ctx, generatorState)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if requeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if generatorState.Spec.GarbageCollectionDeadline != nil {
+		if generatorState.Spec.GarbageCollectionDeadline.Time.Before(time.Now()) {
+			if err := r.Client.Delete(ctx, generatorState, &client.DeleteOptions{}); err != nil {
+				r.markAsFailed("could not delete GeneratorState", err, generatorState)
+				return ctrl.Result{}, fmt.Errorf("could not delete GeneratorState: %w", err)
+			}
+			r.markSuccess("Reached gc deadline: deletion timestamp set", generatorState)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{
+			RequeueAfter: time.Until(generatorState.Spec.GarbageCollectionDeadline.Time),
+		}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) handleFinalizer(ctx context.Context, generatorState *genv1alpha1.GeneratorState) (bool, error) {
+	if generatorState.ObjectMeta.DeletionTimestamp.IsZero() {
+		if added := controllerutil.AddFinalizer(generatorState, generatorStateFinalizer); added {
+			if err := r.Client.Update(ctx, generatorState, &client.UpdateOptions{}); err != nil {
+				return false, fmt.Errorf("could not update finalizers: %w", err)
+			}
+			return true, nil
+		}
+	} else if controllerutil.ContainsFinalizer(generatorState, generatorStateFinalizer) {
+		gen, err := r.getGenerator(generatorState.Spec.Resource.Raw)
+		if err != nil {
+			r.markAsFailed("could not get generator", err, generatorState)
+			return false, fmt.Errorf("could not get generator: %w", err)
+		}
+
+		if err := gen.Cleanup(ctx, generatorState.Spec.Resource, generatorState.Spec.State, r.Client, generatorState.Namespace); err != nil {
+			r.markAsFailed("could not cleanup generator state", err, generatorState)
+			return false, fmt.Errorf("could not cleanup generator state: %w", err)
+		}
+
+		controllerutil.RemoveFinalizer(generatorState, generatorStateFinalizer)
+		if err := r.Client.Update(ctx, generatorState, &client.UpdateOptions{}); err != nil {
+			return false, fmt.Errorf("could not update finalizers: %w", err)
+		}
+	}
+	return false, nil
+}
+
+func (r *Reconciler) getGenerator(resource []byte) (genv1alpha1.Generator, error) {
+	us := &unstructured.Unstructured{}
+	if err := us.UnmarshalJSON(resource); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal resource: %w", err)
+	}
+	gen, ok := genv1alpha1.GetGeneratorByName(us.GroupVersionKind().Kind)
+	if !ok {
+		return nil, fmt.Errorf("generator not found")
+	}
+	return gen, nil
+}
+
+func (r *Reconciler) markAsFailed(msg string, err error, gs *genv1alpha1.GeneratorState) {
+	r.recorder.Event(gs, v1.EventTypeWarning, "Error", fmt.Sprintf("%s: %v", msg, err))
```

Comment:
Shouldn't this do something else as well? Like set a condition or something?

Reply:
Especially based on the changes in https://github.com/external-secrets/external-secrets/pull/4317, we should no longer use events to signal what is wrong with a resource.

This is mainly because the same event gets raised every failed reconcile, whereas the condition message only gets set one time, and are more visible to the end user either way.